File: app/src/main/java/knf/kuma/explorer/DownloadingAdapter.java
package knf.kuma.explorer;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;

import java.util.List;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.database.CacheDB;
import knf.kuma.database.dao.DownloadsDAO;
import knf.kuma.download.DownloadManager;
import knf.kuma.pojos.DownloadObject;
import me.zhanghai.android.materialprogressbar.MaterialProgressBar;

public class DownloadingAdapter extends RecyclerView.Adapter<DownloadingAdapter.DownloadingItem> {

    private Fragment fragment;
    private DownloadsDAO downloadsDAO = CacheDB.INSTANCE.downloadsDAO();
    private List<DownloadObject> downloadObjects;

    DownloadingAdapter(Fragment fragment, List<DownloadObject> downloadObjects) {
        this.fragment = fragment;
        this.downloadObjects = downloadObjects;
    }

    @NonNull
    @Override
    public DownloadingItem onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new DownloadingItem(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.item_downloading_extra, viewGroup, false));
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(@NonNull final DownloadingItem holder, int position) {
        final DownloadObject downloadObject = downloadObjects.get(position);
        holder.title.setText(downloadObject.name);
        holder.chapter.setText(downloadObject.chapter);
        holder.eta.setText(downloadObject.getSubtext());
        holder.progress.setMax(100);
        if (downloadObject.state == DownloadObject.PENDING) {
            holder.eta.setVisibility(View.GONE);
            holder.progress.setIndeterminate(true);
            holder.progress.setProgress(0);
        } else {
            if (downloadObject.state == DownloadObject.PAUSED)
                holder.eta.setVisibility(View.GONE);
            else
                holder.eta.setVisibility(View.VISIBLE);
            holder.progress.setIndeterminate(false);
            holder.progress.setProgress(downloadObject.progress);
        }
        holder.action.setOnClickListener(view -> {
            if (downloadObject.state == DownloadObject.DOWNLOADING) {
                downloadObject.state = DownloadObject.PAUSED;
                holder.action.setText("REANUDAR");
                DownloadManager.pause(downloadObject);
            } else if (downloadObject.state == DownloadObject.PAUSED) {
                downloadObject.state = DownloadObject.PENDING;
                holder.action.setText("PAUSAR");
                DownloadManager.resume(downloadObject);
            }
        });
        holder.cancel.setOnClickListener(v -> {
            try {
                new MaterialDialog.Builder(Objects.requireNonNull(fragment.getContext()))
                        .content("¿Cancelar descarga del " + downloadObject.chapter.toLowerCase() + " de " + downloadObject.name + "?")
                        .positiveText("CONFIRMAR")
                        .negativeText("CANCELAR")
                        .onPositive((dialog, which) -> {
                            downloadObjects.remove(holder.getAdapterPosition());
                            notifyItemRemoved(holder.getAdapterPosition());
                            //downloadsDAO.deleteByEid(downloadObject.eid);
                            DownloadManager.cancel(downloadObject.eid);
                        }).build().show();
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
        downloadsDAO.getLiveByKey(downloadObject.key).observe(fragment, object -> {
            try {
                if (object == null || object.state == DownloadObject.COMPLETED) {
                    downloadObjects.remove(holder.getAdapterPosition());
                    notifyItemRemoved(holder.getAdapterPosition());
                } else {
                    downloadObject.state = object.state;
                    if (object.state == DownloadObject.PENDING) {
                        holder.eta.setVisibility(View.GONE);
                        holder.progress.setIndeterminate(true);
                        holder.progress.setProgress(0);
                    } else {
                        switch (downloadObject.state) {
                            case DownloadObject.DOWNLOADING:
                                holder.action.setText("PAUSAR");
                                holder.eta.setVisibility(View.VISIBLE);
                                break;
                            case DownloadObject.PAUSED:
                                holder.action.setText("REANUDAR");
                                holder.eta.setVisibility(View.GONE);
                                break;
                        }
                        holder.progress.setIndeterminate(false);
                        holder.progress.setProgress(object.progress);
                        holder.eta.setText(object.getSubtext());
                    }
                }
            } catch (Exception e) {
                //
            }
        });
    }

    @Override
    public int getItemCount() {
        return downloadObjects.size();
    }

    class DownloadingItem extends RecyclerView.ViewHolder {
        @BindView(R.id.title)
        TextView title;
        @BindView(R.id.chapter)
        TextView chapter;
        @BindView(R.id.eta)
        TextView eta;
        @BindView(R.id.action)
        Button action;
        @BindView(R.id.cancel)
        Button cancel;
        @BindView(R.id.progress)
        MaterialProgressBar progress;

        DownloadingItem(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/knf/kuma/explorer/DownloadingAdapter.kt
package knf.kuma.explorer

import android.annotation.SuppressLint
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ProgressBar
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import com.afollestad.materialdialogs.MaterialDialog
import knf.kuma.R
import knf.kuma.commons.PrefsUtil
import knf.kuma.commons.doOnUI
import knf.kuma.commons.safeShow
import knf.kuma.database.CacheDB
import knf.kuma.download.DownloadManager
import knf.kuma.pojos.DownloadObject
import org.jetbrains.anko.find
import java.util.*

class DownloadingAdapter internal constructor(private val fragment: Fragment, private val downloadObjects: MutableList<DownloadObject>) : RecyclerView.Adapter<DownloadingAdapter.DownloadingItem>() {
    private val downloadsDAO = CacheDB.INSTANCE.downloadsDAO()

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): DownloadingItem {
        return DownloadingItem(LayoutInflater.from(viewGroup.context).inflate(R.layout.item_downloading_extra, viewGroup, false))
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: DownloadingItem, position: Int) {
        val downloadObject = downloadObjects[position]
        holder.server.text = downloadObject.downloadServer
        holder.title.text = downloadObject.name
        holder.chapter.text = downloadObject.chapter
        holder.eta.text = downloadObject.subtext
        holder.progress.max = 100
        holder.action.visibility = if (downloadObject.canResume) View.VISIBLE else View.INVISIBLE
        if (downloadObject.state == DownloadObject.PENDING) {
            holder.eta.visibility = View.GONE
            holder.progress.isIndeterminate = true
            holder.progress.progress = 0
        } else {
            if (downloadObject.state == DownloadObject.PAUSED)
                holder.eta.visibility = View.GONE
            else
                holder.eta.visibility = View.VISIBLE
            holder.progress.isIndeterminate = false
            holder.progress.progress = downloadObject.progress
        }
        holder.action.setOnClickListener {
            if (downloadObject.state == DownloadObject.DOWNLOADING) {
                downloadObject.state = DownloadObject.PAUSED
                holder.action.text = "REANUDAR"
                DownloadManager.pause(downloadObject)
            } else if (downloadObject.state == DownloadObject.PAUSED) {
                downloadObject.state = DownloadObject.PENDING
                holder.action.text = "PAUSAR"
                DownloadManager.resume(downloadObject)
            }
        }
        holder.cancel.setOnClickListener {
            fragment.context?.let {
                MaterialDialog(it).safeShow {
                    message(text = "¿Cancelar descarga del ${downloadObject.chapter.lowercase(Locale.getDefault())} de ${downloadObject.name}?")
                    positiveButton(text = "CONFIRMAR") {
                        try {
                            downloadObjects.removeAt(holder.adapterPosition)
                            notifyItemRemoved(holder.adapterPosition)
                            DownloadManager.cancel(downloadObject.eid)
                        } catch (e: Exception) {
                            //
                        }
                    }
                    negativeButton(text = "CANCELAR")
                }
            }
        }
        downloadsDAO.getLiveByKey(downloadObject.key).observe(fragment, Observer { downloadObject1 ->
            try {
                if (downloadObject1 == null || downloadObject1.state == DownloadObject.COMPLETED) {
                    downloadObjects.removeAt(holder.adapterPosition)
                    notifyItemRemoved(holder.adapterPosition)
                } else {
                    downloadObject.state = downloadObject1.state
                    if (downloadObject1.state == DownloadObject.PENDING) {
                        holder.eta.visibility = View.GONE
                        holder.progress.isIndeterminate = true
                        holder.progress.progress = 0
                    } else {
                        when (downloadObject.state) {
                            DownloadObject.DOWNLOADING -> {
                                holder.action.text = "PAUSAR"
                                holder.eta.visibility = View.VISIBLE
                            }
                            DownloadObject.PAUSED -> {
                                holder.action.text = "REANUDAR"
                                holder.eta.visibility = View.GONE
                            }
                        }
                        holder.progress.isIndeterminate = false
                        if (downloadObject1.getEta() == -2L || PrefsUtil.downloaderType == 0)
                            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
                                holder.progress.setProgress(downloadObject1.progress, true)
                            else
                                holder.progress.progress = downloadObject1.progress
                        else {
                            holder.progress.progress = 0
                            holder.progress.secondaryProgress = downloadObject1.progress
                        }
                        holder.eta.text = downloadObject1.subtext
                    }
                }
            } catch (e: Exception) {
                //
            }
        })
    }

    override fun getItemCount(): Int {
        return downloadObjects.size
    }

    fun remove(eid: String) {
        ArrayList(downloadObjects).forEachIndexed { index, downloadObject ->
            if (downloadObject.eid == eid) {
                downloadObjects.removeAt(index)
                fragment.doOnUI { notifyItemRemoved(index) }
                return
            }
        }
    }

    inner class DownloadingItem(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val server: TextView = itemView.find(R.id.server)
        val title: TextView = itemView.find(R.id.title)
        val chapter: TextView = itemView.find(R.id.chapter)
        val eta: TextView = itemView.find(R.id.eta)
        val action: Button = itemView.find(R.id.action)
        val cancel: Button = itemView.find(R.id.cancel)
        val progress: ProgressBar = itemView.find(R.id.progress)
    }
}
